### PR TITLE
[SYCL-PTX] Add math builtins categories

### PR DESCRIPTION
**Pull request by Naghasan**
The patch adds the generic SPIR-V implementation for the math builtins (so nvptx specific implementation yet).

Signed-off-by: Victor Lomuller <victor@codeplay.com>
Co-authored-by: David Wood <david.wood@codeplay.com>

### CODE REVIEW

File: libclc/generic/libspirv/math/nan.inc [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#ifdef __CLC_SCALAR
```
Comment (bader) [inline libclc/generic/libspirv/math/nan.inc:1]:
```suggestion
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
#ifdef __CLC_SCALAR
```

File: libclc/generic/libspirv/math/clc_sw_unary.inc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#include <utils.h>
```
Comment (bader) [inline libclc/generic/libspirv/math/clc_sw_unary.inc:1]:
```suggestion
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
#include <utils.h>
```

File: libclc/generic/libspirv/math/clc_sw_binary.inc [hunk 3]
```diff
@@ -1,9 +1,6 @@
 #include <utils.h>
```
Comment (bader) [inline libclc/generic/libspirv/math/clc_sw_binary.inc:1]:
```suggestion
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
#include <utils.h>
```

File: libclc/generic/libspirv/math/asin.inc [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// TODO: Enable half precision when atan2 is implemented
```
Comment (bader) [inline libclc/generic/libspirv/math/asin.inc:1]:
```suggestion
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
// TODO: Enable half precision when atan2 is implemented
```

File: libclc/generic/libspirv/math/clc_hypot.cl [hunk 5]
```diff
@@ -1,31 +1,17 @@
...
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
- * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
- * THE SOFTWARE.
- */
-
-#include <clc/clc.h>
+//===----------------------------------------------------------------------===//
```
Comment (bader) [inline libclc/generic/libspirv/math/clc_hypot.cl:1]: Hm... shouldn't we keep old AMD copyright header?

File: libclc/generic/libspirv/common/degrees.cl [hunk 6]
```diff
@@ -8,7 +8,7 @@
 
 #include <spirv/spirv.h>
 
-#include "../../lib/clcmacro.h"
+#include <clcmacro.h>
```
[the remark this answers is not in the document]
Comment (bader) [reply libclc/generic/libspirv/common/degrees.cl:11]: I assume these "common" built-ins are used in "math" built-ins implementation, right?

[the remark this answers is not in the document]
Comment (Naghasan) [reply libclc/generic/libspirv/common/degrees.cl:11]: > This probably should belong to other PR.

That's a change that relates to the refactoring prior to this patch. This wasn't part of it as it directly conflicted with the PR https://github.com/intel/llvm/pull/1832. This avoided a dependency.

The header file included is the same, `../../lib/clcmacro.h` is a link to the one in `include`. So there is no functional change, just consistency.

> I assume these "common" built-ins are used in "math" built-ins implementation, right?

Some `common` builtins are indeed used by the `math` one.

File: libclc/generic/libspirv/math/clc_hypot.cl (hunk 5 above)
Comment (Naghasan) [reply libclc/generic/libspirv/math/clc_hypot.cl:1]: I can revert if you prefer.

Technically, the LLVM relicensed all sub-projects under this license, but this one never updated the headers. But I'm no lawyer.

**Pull request merged**